python/tchannel/outgoing.py

```diff
@@ -0,0 +1,342 @@
+
+        No-op if the handshake was already performed.
+        """
+        assert (
+            not self._state == self.State.closed
+        ), "The connection to %s:%s has already been terminated." % (
+            self.host, self.port
+        )
+
+        if self._state != self.State.init:
+            return
+
+        self._conn.initiate_handshake({
+            'host_port': '0.0.0.0:0',  # We can't receive requests
+            'process_name': self._name
+        })
+        resp = self._conn.await_handshake_reply()
+        self._state = self.State.ready
+        return resp
+
+    def send(self, arg_1, arg_2, arg_3, service=None, async=False):
+        """Send the given arguments over the wire.
+
+        ``arg_1``, ``arg_2``, and ``arg_3`` represent the triple being sent
+        over the wire.
+
+        `service` is the name of the service being called. It defaults to an
+        empty string.
+
+        :param async:
+            Whether this should block for the response. The function returns a
+            Future if True.
+        :raises TChannelApplicationException:
+            If the remote server did not respond with status code 0.
+        """
+        assert (
+            self._state == self.State.ready
+        ), "Handshake not performed or connection closed"
+
+        msg = CallRequestMessage(arg_1=arg_1, arg_2=arg_2, arg_3=arg_3)
+
+        future = SettableFuture()
+        self._submit(msg, future)
+
+        def handle_response(response):
+            if response.code != 0:
+                raise TChannelApplicationException(
+                    response.code,
+                    response.arg_1,
+                    response.arg_2,
+                    response.arg_3,
+                )
+            return response
+
+        result_future = transform_future(future, handle_response)
+        if async:
+            return result_future
+        else:
+            # Wait till the response for this message has been received.
+            return result_future.result()  # TODO: timeout?
+
+    def closed(self):
+        """Returns True if the connection has been closed."""
+        return self._state == self.State.closed
+
+    def close(self, wait=True):
+        """Manually close this connection
+
+        It is generally not necessary to call this manually because the
+        TChannel instance will clean up after itself.
+        """
+        if self.closed():
+            return
+        self._state = self.State.closed
+
+        # A None request informs the sender it's time to go.
+        self._outstanding.put(None)
+        self._sender.join()
+
+        # Closing the socket should stop the receiver waiting.
+        self._sock.close()
+        self._receiver.join()
+
+        self._on_close(self)
+
+    def __del__(self):
+        try:
+            self.close()
+        except:
+            pass
+
+
+class OutgoingTChannel(object):
```

Comment:
Both `TChannel` interfaces support outgoing so I wonder if we could call this something more descriptive. Otherwise lgtm, I look forward to the merge conflicts :)